add support for BEnumOverride parameter for override action invocation - issue #44

naimakov opened the pull request:
A numeric ordinal or string tag values are allowed in the invokeAction request

nurkanat-aimakov-fr committed:
add construction of argument for BEnumOverride action

A numeric ordinal or string tag values are allowed in the invokeAction request

ci-richard-mcelhinney commented on nhaystack-rt/src/nhaystack/util/TypeUtil.java:
```diff
@@ -226,6 +228,46 @@ else if (def instanceof BSimple)
+            BEnumRange range = (BEnumRange) facets.get(BFacets.RANGE);
+
+            Iterator<Map.Entry<String, HVal>> it = args.iterator();
+            while (it.hasNext())
```
What is the reason to loop here?  Are you expecting more than 1 argument to be present?

naimakov replied:
Yes, duration and maxOverrideDuration

ci-richard-mcelhinney commented on nhaystack-rt/src/nhaystack/util/TypeUtil.java:
```diff
@@ -226,6 +228,46 @@ else if (def instanceof BSimple)
+            {
+                Map.Entry<String, HVal> e = it.next();
+                BSimple value = toBajaSimple(e.getValue());
+                if (!e.getKey().equals("value"))
```
In a BEnumOverride there are only 3 possible fields you can set, duration, value, and maxOverrideDuration.  In this 'if' block what would happen if another key not equals to 'value' and also not one of the 3 permitted fields was passed in?

naimakov replied:
That's true, it is assumed that the user provides only valid inputs. I can rewrite the code to only allow for 3 permitted fields

ci-richard-mcelhinney commented on nhaystack-rt/src/nhaystack/util/TypeUtil.java:
```diff
@@ -226,6 +228,46 @@ else if (def instanceof BSimple)
+        {
+            BEnumOverride cpx = (BEnumOverride) def;
+            BFacets facets = comp.getAction(action.getName()).getFacets();
+            BEnumRange range = (BEnumRange) facets.get(BFacets.RANGE);
```
What happens if there is not a BEnumRange configured in the facets for the EnumWritable? Does 'range' get set to null?

naimakov replied:
True, I should guard against wrong comp + action combination here

ci-richard-mcelhinney commented on nhaystack-rt/src/nhaystack/util/TypeUtil.java:
```diff
@@ -226,6 +228,46 @@ else if (def instanceof BSimple)
+                // construction of BDynamicEnum arg
+                if (value instanceof BString && range.isTag(value.toString()))
+                {
+                    value = range.get(value.toString());
```
Referring to line 235 above and also my comment, if 'range' is not set on line 235 to a valid BEnumRange, is it null here at line 250, could this possibly throw an exception?

naimakov replied:
Yes, it could

ci-richard-mcelhinney commented on nhaystack-rt/src/nhaystack/util/TypeUtil.java:
```diff
@@ -226,6 +228,46 @@ else if (def instanceof BSimple)
+                }
+                else if (value instanceof BDouble && range.isOrdinal(((BDouble) value).getInt()))
+                {
+                    value = range.get(((BDouble) value).getInt());
```
Same issue here as described above for line 250.

ci-richard-mcelhinney reviewed (commented):
@naimakov thank you very much for your PR and contribution.  I have made a couple of comments.

Is it possible for you to add some tests into cover the changes you are proposing?

ci-richard-mcelhinney commented:
@naimakov did you see my comments on your PR?  Can you either update or provide some responses please?

naimakov commented:
Yes, I can add some tests

naimakov commented:
Thank you for your review and valuable feedback, @ci-richard-mcelhinney.
I'll take feedback and modify the PR

ci-richard-mcelhinney commented:
@naimakov no problem.  I will review any updated code as soon as you post it.

Thanks,
Richard

nurkanat-aimakov-fr committed:
add facets validation

ci-richard-mcelhinney commented:
@naimakov have you provided any updates yet for me to review?

nurkanat-aimakov-fr committed:
add arg types validation

naimakov commented:
> @naimakov have you provided any updates yet for me to review?

@ci-richard-mcelhinney hi, sorry for the long wait.

I refactored action argument iteration and added checks against invalid args

naimakov commented:
@ci-richard-mcelhinney I'm adding unit tests, let me come back to you later today.

nurkanat-aimakov-fr committed:
add unit tests

naimakov commented:
@ci-richard-mcelhinney thank you for valuable feedback and patience. I've added modifications and unit tests for the changes proposed.

naimakov commented:
Hi, @ci-richard-mcelhinney. I wanted to follow up on my previous comment. Would you be able to review PR changes I made please? Thank you.

nurkanat-aimakov-fr committed:
remove commented-out code
